tests/shared/src/test/scala/munit/AsyncFixtureSuite.scala

```diff
@@ -0,0 +1,113 @@
+package munit
+
+import scala.concurrent.Future
+import scala.concurrent.Promise
+
+class AsyncFixtureSetupTestTeardownOrderSuite extends FunSuite {
+  val latch = Promise[Unit]
+  var completedFromTest = Option.empty[Boolean]
+  var completedFromTeardown = Option.empty[Boolean]
+
+  val latchOnTeardown = FunFixture.async[String](
+    setup = { test => Future.successful(test.name) },
+    teardown = { name =>
+      implicit val ec = munitExecutionContext
+      Future {
+        completedFromTeardown = Some(latch.trySuccess(()));
+      }
+    }
+  )
+
+  override def afterAll(): Unit = {
+    assertEquals(completedFromTest, Some(true))
+    assertEquals(completedFromTeardown, Some(false))
+  }
+
+  latchOnTeardown.test("test is ran before the teardown") { _ =>
+    import scala.concurrent.ExecutionContext.Implicits.global
+    Future {
+      // Ideally we would simulate some work here, for example by using Thread.sleep(50),
+      // which would increase the certainty that this test passes by design and not by lucky
+      // scheduling. However Thread.sleep is not scala-native compatible.
+      //Thread.sleep(50)
+      completedFromTest = Some(latch.trySuccess(()))
+    }
+  }
+}
+
+class AsyncFixtureTeardownSuite extends FunSuite {
+  @volatile var cleanedUp: Boolean = _
+
+  val cleanupInTeardown = FunFixture.async[Unit](
+    _ => { cleanedUp = false; Future.successful(()) },
+    _ => { cleanedUp = true; Future.successful(()) }
+  )
+
+  override def afterAll(): Unit = {
+    assert(cleanedUp)
+  }
+
+  cleanupInTeardown.test("calls teardown when test throws".fail) { _ =>
+    throw new Error("failure in test")
+  }
+
+  cleanupInTeardown.test("calls teardown when test returns failed Future".fail) {
+    _ => Future.failed(new Error("failure in test"))
+  }
+}
+
+class AsyncFixtureSuite extends FunSuite {
```

Comment:
Great test suite! It's nice to cover combinations of failing setup/teardown fixutres.

I think it would be helpful to convert this into a "framework suite" instead of using `.fail`. Framework suites allow us to see the culprit exception message that caused the failure

Reply:
Thanks for pointing me to FrameworkSuite. Its quite awesome. I converted my suite to it and that found a bug in my implementation!